history.go

```diff
@@ -0,0 +1,75 @@
+package merkletree
+
+import (
+	"errors"
+
+	"github.com/coniks-sys/libmerkleprefixtree-go/crypto"
+)
+
+var (
+	ErrHistoryExisted = errors.New("[merkletree] trying to override existing history")
+	ErrBadEpoch       = errors.New("[merkletree] next epoch's STR has bad epoch ")
+)
+
+type History struct {
+	tree          *MerkleTree
+	currentSTR    *SignedTreeRoot
+	epochInterval int64
+}
+
+func NewHistory(m *MerkleTree, key crypto.SigningKey, startEp, epInterval int64) *History {
+	h := new(History)
+	h.tree = m
+	h.epochInterval = epInterval
+	h.currentSTR = newSTR(m, startEp, 0, make([]byte, crypto.HashSizeByte), key)
+	return h
+}
+
+func (h *History) UpdateHistory(m *MerkleTree, nextEp int64) error {
```

Comment:
> This way, new data would be inserted to the tree at the end of epoch, not the next version of the tree. This is different from what is proposed in the paper (section 4.1.1)

> It also rises another issue: we have 2 different STRs of each epoch, right? One from the beginning of the epoch to before the end of the epoch, and the other is at the end of the epoch. Did I missed something here?

I may be reading this wrong, but as I understand it, an epoch is discrete. The current epoch is really the most recent snapshot that's been captured.  Remember, in between epochs (including the first) we hand out temporary bindings, which aren't yet included in the str (but are find to put in the tree since they won't be included in the current epoch until the next snapshot occurs).

> Yes. I mean there are some extra works need to be done by the developer before/after updating the tree. Of course we can setup the timer as a part of the lib but I think it would have no advantages here. I need to think more about it.

Ok, good. I was really asking because I don't know either. You make a good point.